File: TypeScript/ExampleTests/common/FilterValueTyped.ts
import { FilterValueString } from "./FilterValueString.js";
import * as _json from "./json.js";

export class FilterValueTyped {
  value: IDForm;

  constructor(value: IDForm) {
    this.value = value;
  }

  static fromStringObj(s: FilterValueString): FilterValueTyped {
    return new FilterValueTyped(
      new IDForm(s.value)
    );
  }

  toJsonValue(): Record<string, unknown> {
    return {
      value: this.value == null ? null : String(this.value),
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): FilterValueTyped {
    return new FilterValueTyped(
      new IDForm(_json.asString(_json.requireField(m, "value"), "value"))
    );
  }

  static fromJSON(text: string): FilterValueTyped {
    return FilterValueTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly FilterValueTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): FilterValueTyped[] {
    const raw = _json.asArray(_json.parse(text), "FilterValueTyped") ?? [];
    return raw.map((e) => FilterValueTyped.fromJsonValue(e));
  }
}

File: Rust/ExampleTests/common/filtervalue_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::filtervalue_string::FilterValueString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct FilterValueTyped {
    pub value: String,
}

impl FilterValueTyped {
    pub fn from_str_struct(s: &FilterValueString) -> Self {
        Self {
            value: s.value.clone(),
        }
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("value".to_string(), json::Value::Str(self.value.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            value: json::as_string(json::require(v, "value")?, "value")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[FilterValueTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<FilterValueTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "FilterValueTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
